flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/SubtaskCheckpointCoordinatorImpl.java

```diff
@@ -527,20 +552,22 @@ private void finishAndReportAsync(
             CheckpointMetaData metadata,
             CheckpointMetricsBuilder metrics,
             Supplier<Boolean> isRunning) {
-        // we are transferring ownership over snapshotInProgressList for cleanup to the thread,
-        // active on submit
-        asyncOperationsThreadPool.execute(
+        AsyncCheckpointRunnable asyncCheckpointRunnable =
                 new AsyncCheckpointRunnable(
                         snapshotFutures,
                         metadata,
                         metrics,
                         System.nanoTime(),
                         taskName,
-                        registerConsumer(),
                         unregisterConsumer(),
                         env,
                         asyncExceptionHandler,
-                        isRunning));
+                        isRunning);
+        registerConsumer().accept(asyncCheckpointRunnable);
```

Comment:
BTW, sorry I missed one comment in the previous review (not sure how). Let's inline the regsiterConsumer() and remove the method. It is unnecessary to go through the Consumer here.

I believe we can replace it with:
```
registerAsyncCheckpointRunnable(
                        asyncCheckpointRunnable.getCheckpointId(), asyncCheckpointRunnable);
```